pkg/cmd/cli/cmd/idle.go

```diff
@@ -0,0 +1,630 @@
+	oldTargetsRaw, hasOldTargets := annotations[unidlingapi.UnidleTargetAnnotation]
+
+	scaleRefs := make([]unidlingapi.RecordedScaleReference, 0, len(rawScaleRefs))
+
+	// initialize the list of new annotations
+	for rawScaleRef := range rawScaleRefs {
+		scaleRefs = append(scaleRefs, unidlingapi.RecordedScaleReference{
+			CrossGroupObjectReference: rawScaleRef,
+			Replicas:                  0,
+		})
+	}
+
+	// if the new preserved scale would be 0, see if we have an old scale that we can use instead
+	if hasOldTargets {
+		var oldTargets []unidlingapi.RecordedScaleReference
+		oldTargetsSet := make(map[unidlingapi.CrossGroupObjectReference]int)
+		if err := json.Unmarshal([]byte(oldTargetsRaw), &oldTargets); err != nil {
+			return nil, fmt.Errorf("unable to extract existing scale information from endpoints %s: %v", serviceName.String(), err)
+		}
+
+		for i, target := range oldTargets {
+			oldTargetsSet[target.CrossGroupObjectReference] = i
+		}
+
+		// figure out which new targets were already present...
+		for _, newScaleRef := range scaleRefs {
+			if oldTargetInd, ok := oldTargetsSet[newScaleRef.CrossGroupObjectReference]; ok {
+				if newScale, ok := scales[newScaleRef.CrossGroupObjectReference]; !ok || newScale == 0 {
+					scales[newScaleRef.CrossGroupObjectReference] = oldTargets[oldTargetInd].Replicas
+				}
+				delete(oldTargetsSet, newScaleRef.CrossGroupObjectReference)
+			}
+		}
+
+		// ...and add in any existing targets not already on the new list to the new list
+		for _, ind := range oldTargetsSet {
+			scaleRefs = append(scaleRefs, oldTargets[ind])
+		}
+	}
+
+	for i := range scaleRefs {
+		scaleRef := &scaleRefs[i]
+		newScale, ok := scales[scaleRef.CrossGroupObjectReference]
+		if !ok || newScale == 0 {
+			newScale = 1
+			if scaleRef.Replicas != 0 {
+				newScale = scaleRef.Replicas
+			}
+		}
+
+		scaleRef.Replicas = newScale
+	}
+
+	return scaleRefs, nil
+}
+
+// setIdleAnnotations sets the given annotation on the given object to the marshaled list of CrossGroupObjectReferences
+func setIdleAnnotations(serviceName types.NamespacedName, annotations map[string]string, scaleRefs []unidlingapi.RecordedScaleReference, nowTime time.Time) error {
+	var scaleRefsBytes []byte
+	var err error
+	if scaleRefsBytes, err = json.Marshal(scaleRefs); err != nil {
+		return err
+	}
+
+	annotations[unidlingapi.UnidleTargetAnnotation] = string(scaleRefsBytes)
+	annotations[unidlingapi.IdledAtAnnotation] = nowTime.Format(time.RFC3339)
+
+	return nil
+}
+
+// patchObj patches calculates a patch between the given new object and the existing marshaled object
+func patchObj(obj runtime.Object, metadata meta.Object, oldData []byte, mapping *meta.RESTMapping, f *clientcmd.Factory) (runtime.Object, error) {
```

Comment:
Actually, I think this was just a wrapper around a couple lines that are just inline elsewhere in the code base, because I needed to do the same thing multiple times at one point.  IMO, it's a bit clearer like this, even though it's just used one place.